File: runtime-go/domain/task_graph.go
package domain

import "sort"

// TaskGraph 是 SessionView.Tasks 上的只读图视图(ch07)。
// 不单独持久化——由 Fold 出的 Tasks 派生。
type TaskGraph struct {
	Roots    []string            // ParentID == "" 的 Task ID
	Children map[string][]string // parentID -> child IDs(按 ID 稳定排序)
}

// BuildTaskGraph 从 SessionView.Tasks 派生图。
func BuildTaskGraph(tasks map[string]Task) TaskGraph {
	g := TaskGraph{
		Children: make(map[string][]string),
	}
	for id, t := range tasks {
		if t.ParentID == "" {
			g.Roots = append(g.Roots, id)
			continue
		}
		g.Children[t.ParentID] = append(g.Children[t.ParentID], id)
	}
	sort.Strings(g.Roots)
	for parentID := range g.Children {
		sort.Strings(g.Children[parentID])
	}
	return g
}

// ChildrenOf 返回直接子 Task ID 列表(可能为空)。
func (g TaskGraph) ChildrenOf(parentID string) []string {
	return g.Children[parentID]
}

// IsLeaf 判断 task 是否没有子节点。
func (g TaskGraph) IsLeaf(taskID string) bool {
	return len(g.Children[taskID]) == 0
}

File: runtime-rs/src/domain/task_graph.rs
//! TaskGraph —— 与 `runtime-go/domain/task_graph.go` 对齐。

use std::collections::HashMap;

use super::Task;

#[derive(Debug, Clone, Default)]
pub struct TaskGraph {
    pub roots: Vec<String>,
    pub children: HashMap<String, Vec<String>>,
}

pub fn build_task_graph(tasks: &HashMap<String, Task>) -> TaskGraph {
    let mut g = TaskGraph::default();
    for (id, t) in tasks {
        if t.parent_id.is_empty() {
            g.roots.push(id.clone());
        } else {
            g.children
                .entry(t.parent_id.clone())
                .or_default()
                .push(id.clone());
        }
    }
    g
}

impl TaskGraph {
    pub fn children_of(&self, parent_id: &str) -> &[String] {
        self.children
            .get(parent_id)
            .map(|v| v.as_slice())
            .unwrap_or(&[])
    }
}
